src/FrameBuffer.cpp

```diff
@@ -1281,18 +1282,126 @@ void _writeToRdram(TSrc* _src, TDst* _dst, TDst(*converter)(TSrc _c), TSrc _test
 			++numStored;
 		}
 		++y;
-		_dst += numStored;
 	}
 
-	u32 dsty = 0;
 	for (; y < _height; ++y) {
 		for (u32 x = 0; x < _width && numStored < _numPixels; ++x) {
 			c = _src[x + (_height - y - 1)*_width];
 			if (c != _testValue)
-				_dst[(x + dsty*_width) ^ _xor] = converter(c);
+				_dst[(x + y*_width) ^ _xor] = converter(c);
 			++numStored;
 		}
-		++dsty;
+	}
+}
+
+template <typename TSrc>
+void _writeToRdram16(TSrc* _src, u16* _dst, u16(*converter)(TSrc _c), TSrc _testValue, u32 _width, u32 _height, u32 _numPixels, u32 _startAddress, u32 _bufferAddress, u32 _bufferSize)
+{
+	u32 chunkStart = ((_startAddress - _bufferAddress) >> (_bufferSize - 1)) % _width;
+	if (chunkStart % 2 != 0) {
+		--chunkStart;
+		--_dst;
+		++_numPixels;
+	}
+
+	u32 numStored = 0;
+	u32 y = 0;
+	TSrc c1, c2;
+	u16 val1, val2;
+
+	u32* _dst32 = reinterpret_cast<u32*>(_dst);
+    u32 dst32index = 0;
+    u32 heightOffset;
+
+	if (chunkStart > 0) {
+		for (u32 x = chunkStart; x < _width; x += 2) {
+		    heightOffset = _height - y - 1;
+			c1 = _src[x + heightOffset*_width];
+		    c2 = _src[x + 1 + heightOffset*_width];
+		    val1 = c1 != _testValue ? converter(c1) : _dst[numStored + 1];
+		    val2 = c2 != _testValue ? converter(c2) : _dst[numStored];
+
+            _dst32[dst32index] = (val1 << 16) | val2;
+
+			numStored += 2;
+			++dst32index;
+		}
+		++y;
+	}
+
+	for (; y < _height; ++y) {
+		for (u32 x = 0; x < _width && numStored < _numPixels; x += 2) {
+		    heightOffset = _height - y - 1;
+			c1 = _src[x + heightOffset*_width];
+		    c2 = _src[x + 1 + heightOffset*_width];
+		    val1 = c1 != _testValue ? converter(c1) : _dst[numStored + 1];
+		    val2 = c2 != _testValue ? converter(c2) : _dst[numStored];
+
+            _dst32[dst32index] = (val1 << 16) | val2;
```

Comment:
I'm hoping I don't have this reversed... I need a game that triggers this to make sure it's right.